### PR TITLE
Diff & merge based on the branch

### PR DESCRIPTION
**Pull request by anandjakhaniya**
Screen Shot 2020-04-02 at 10 01 55 PM

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java [hunk 1]
```diff
@@ -378,22 +379,49 @@ boolean childContains(Set<?> list, Set<?> sublist) {
...
-      CommitEntity internalCommitB = session.get(CommitEntity.class, request.getCommitB());
-      if (internalCommitB == null) {
-        throw new ModelDBException(
-            "No such commit found : " + request.getCommitB(), Status.Code.NOT_FOUND);
+      RepositoryEntity repositoryEntity =
+          repositoryDAO.getRepositoryById(session, request.getRepositoryId());
+
+      CommitEntity internalCommitA;
+      CommitEntity internalCommitB;
+      if (request.getBranchA().isEmpty() || request.getBranchB().isEmpty()) {
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: I can merge commit `24325` into `master`

Comment (ravishetye) [reply backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: extract commitA and commitB on separate ifs

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java [hunk 2]
```diff
@@ -489,18 +530,31 @@ boolean childContains(Set<?> list, Set<?> sublist) {
...
-      RepositoryEntity repositoryEntity = repositoryFunction.apply(readSession);
+      RepositoryEntity repositoryEntity =
+          repositoryDAO.getRepositoryById(readSession, request.getRepositoryId());
 
-      internalCommitA = readSession.get(CommitEntity.class, request.getCommitShaA());
-      if (internalCommitA == null) {
-        throw new ModelDBException(
-            "No such commit found : " + request.getCommitShaA(), Status.Code.NOT_FOUND);
-      }
+      if (request.getBranchA().isEmpty() || request.getBranchB().isEmpty()) {
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: same as commit

**Review by ravishetye — CHANGES_REQUESTED** [review]
minor changes requested , if done can be merge post a quick verification by @evgeniilezhnin

**Comment by ravishetye** [comment]
Once this is in , also generate the protos and merge the branch this is based on

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java (hunk 1 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: Resolved as per attached scenarios image on the description

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java (hunk 2 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: Resolved as per attached scenarios image on the description

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java [hunk 3]
```diff
@@ -378,22 +379,71 @@ boolean childContains(Set<?> list, Set<?> sublist) {
...
+        invalidParam = true;
+      } else if (!request.getBranchA().isEmpty() && !request.getCommitA().isEmpty()) {
+        invalidParam = true;
+      } else if (request.getBranchA().isEmpty() && request.getCommitA().isEmpty()) {
+        invalidParam = true;
+      } else if (!request.getBranchB().isEmpty() && !request.getCommitB().isEmpty()) {
+        invalidParam = true;
+      } else if (request.getBranchB().isEmpty() && request.getCommitB().isEmpty()) {
+        invalidParam = true;
+      }
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]:
```suggestion
      final boolean invalidParam;
      if (request.getBranchA().isEmpty()
          && request.getBranchB().isEmpty()
          && !request.getCommitA().isEmpty()
          && !request.getCommitB().isEmpty()
          || !request.getBranchA().isEmpty()
          && !request.getBranchB().isEmpty()
          && request.getCommitA().isEmpty()
          && request.getCommitB().isEmpty()) {
        invalidParam = false;
      } else  {
        invalidParam = true;
      }
```
maybe this variant is simpler?

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java [hunk 4]
```diff
@@ -479,28 +529,77 @@ boolean childContains(Set<?> list, Set<?> sublist) {
...
+      invalidParam = true;
+    } else if (!request.getBranchA().isEmpty() && !request.getCommitShaA().isEmpty()) {
+      invalidParam = true;
+    } else if (request.getBranchA().isEmpty() && request.getCommitShaA().isEmpty()) {
+      invalidParam = true;
+    } else if (!request.getBranchB().isEmpty() && !request.getCommitShaB().isEmpty()) {
+      invalidParam = true;
+    } else if (request.getBranchB().isEmpty() && request.getCommitShaB().isEmpty()) {
+      invalidParam = true;
+    }
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: if it's the same code as in computeRepositoryDiff (except request type) consider reusing the code

**Review by evgeniilezhnin — APPROVED** [review]
minor comments

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java (hunk 3 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: oh, sorry misunderstood the logic

Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: just ignore this comment

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java (hunk 4 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: fixed

**Pull request merged**